pkg/api/types.go

```diff
@@ -1407,6 +1411,69 @@ type ReplicationControllerList struct {
+type DisruptionBudgetSpec struct {
+	// RateLimits is a list of rate limits on disruptions (preemption and other evictions) across the corresponding
+	// collection of pods, e.g. no more than one disruption per hour across the pods of a particular Service
+	RateLimits []RateLimitSpec `json:"ratelimits,omitempty"`
+	// The minium number of pods that must be available simultaneously.
+	// Value can be an absolute number (ex: 5) or a percentage of total pods(ex: 10%).
+	// Absolute number is calculated from percentage by rounding up.
+	// By default, a fixed value of 1 is used.
+	MinAvailable intstr.IntOrString `json:"minAvailable,omitempty"`
+	// Minimum number of seconds for which a newly created pod should be ready
+	// without any of its container crashing, for it to be considered available.
+	// Defaults to 0 (pod will be considered available as soon as it is ready)
+	MinReadySeconds int `json:"minReadySeconds,omitempty"`
+}
+
+// DisruptionBudgetStatus represents information about the status of a DisruptionBudget. Status may trail the actual
+// state of a system.
+type DisruptionBudgetStatus struct {
+	// The current disruption rate that the collection of pods is experiencing.
+	// This data may be out of date by some window of time.
+	RateLimitStatuses []RateLimitStatus `json:"RateLimitStatuses,omitempty"`
+	// Total number of available pods (ready for at least minReadySeconds) targeted by this disruption budget.
+	AvailableReplicas int `json:"availableReplicas,omitempty"`
+	// Total number of unavailable pods targeted by this disruption budget.
+	UnavailableReplicas int `json:"unavailableReplicas,omitempty"`
+}
+
+// RateLimitSpec describes the rate limit on disruptions (preemption and other evictions) across the corresponding
+// collection of pods
+type RateLimitSpec struct {
+	// Capacity is the maximum number of DisruptionToken could exist during a period of time.
+	// Once there are more than Capacity number of DisruptionToken during a period of time, we can
+	// not disrupt pods in a further step.
+	Capacity inf.Dec `json:"capacity,omitempty"`
+	// Duration is the time unit of RateLimit
+	Duration time.Duration `json:"duration,omitempty"`
+}
+
+// Status is the current status of a ratelimit.
+type RateLimitStatus struct {
+	// ObservedDisruptionTokenNum if the observed number of DisruptionToken
+	ObservedDisruptionTokenNum inf.Dec `json:"observedDisruptionTokenNum,omitempty"`
+	// Duration is the time unit of disruption rate the collection of pods is experiencing.
+	Duration time.Duration `json:"duration,omitempty"`
+}
+
+// DisruptionToken is used to limit the rate of pod disruption.
+// When disruption (preemption and other evictions) happens, we create a DisruptionToken
+// (with DeletionTimeStamp, so that API server could delete it when it expired).
+type DisruptionToken struct {
+	// ServiceOrJobRef is reference to the Service or Job
+	ServiceOrJobRef LocalObjectReference `json:"serviceOrJobRef,omitempty"`
+}
```

Comment:
For the first version, I think we should just have an (optional) pointer from each Pod to its DisruptionBudget. To address @bgrant0607's comment about v1 API, we could make it an annotation on Pod instead of a field of PodSpec. So in sum, you would
* remove the field you added to ServiceSpec
* remove the field you added to JobSpec
* remove this field in DisruptionToken
And then when we actually use the feature (e.g. user creates a Pod they want protected, or node drain tool is considering draining a machine), we would use the Pod annotation.

I think that for the second version, for Job and Service we might want to inline the DisruptionBudgetSpec and have the controller create a corresponding DisruptionBudget object (which it could reflect back in the JobStatus/ServiceStatus), rather than having the user specify a reference to a DisruptionBudget in the JobSpec/ServiceSpec.